added email-password validation and navigation from auth to main (with data passing)

cValentyna opened the pull request:
Added email-password validation:
errors are shown if user pressed registration button or tried to fill textfield but lost focus before correct information was filled in.

Added navigation from auth to main (with data passing), also animation was added between activities

cValentyna committed:
level1 - added email-password validation
added navigation from auth to main (with data passing)

cValentyna committed:
level1 - added animation between activities

1937Carrie commented on app/src/main/java/com/project/myapp/AuthActivity.kt:
```diff
@@ -2,58 +2,203 @@ package com.project.myapp
+     */
+    private fun getName(email: String): String {
+        val partOfEmail =
+            email.substring(0, email.indexOf('@')).replace("[._]".toRegex(), " ").lowercase()
```
```kotlin
val partOfEmail = email.substringBefore("@").split(".", "_").joinToString(" ")
```

cValentyna replied:
It was fixed.

1937Carrie commented on app/src/main/java/com/project/myapp/AuthActivity.kt:
```diff
@@ -2,58 +2,203 @@ package com.project.myapp
     }
+
+    private fun validateEmail(text: CharSequence?) {
+        val pattern = Regex("^[\\w-.]+@([\\w-]+\\.)+\\w{2,}\$")
```
Built-in email address pattern matcher `Patterns.EMAIL_ADDRESS`.
```kotlin
Patterns.EMAIL_ADDRESS.matcher(text.toString()).matches() // check if the email pattern matches the given text
```

cValentyna replied:
Also it was changed.

1937Carrie commented on app/src/main/java/com/project/myapp/AuthActivity.kt:
```diff
@@ -2,58 +2,203 @@ package com.project.myapp
     private val binding: ActivityAuthBinding by lazy {
         ActivityAuthBinding.inflate(layoutInflater)
     }
+    private var isUserEmailValid = false
+    private var isUserPasswordValid = false
+
+    private var wasFocusEmail = false
+    private var firstFocusEmail = false
+    private var wasFocusPassword = false
+    private var firstFocusPassword = false
```
This is looks like properties for holding UI state. If you want to hold it check out this [topic](https://developer.android.com/topic/architecture/ui-layer). If this turns out to be difficult, contact on discord and you will get help.

cValentyna replied:
Unfortunately I couldn't figure it out myself.

cValentyna replied:
Done.

cValentyna committed:
level1 - corrected email pattern, getName()

cValentyna committed:
level1 - added viewModel  for saving UI state.

1937Carrie reviewed (approved):
Overall, it was done very well. If you will need help with providing improves to the project - let me know in discord. What else can be improved:

1. You hold email and password validators right in activity class. Activity belongs to the UI layer, text validators belong to the data layer. Validators can be extracted outside activity class and be calling in viewmodel for passing gotten result of validation to activity;
2. I see you left some comments above classes, functions. There is exists some different type of comments, [source](https://kotlinlang.org/docs/basic-syntax.html#comments):
```kotlin
// This is an end-of-line comment

/* This is a block comment
   on multiple lines. */

/* The comment starts here
/* contains a nested comment */
and ends here. */

// Below is KDoc's comment
/**
 * A group of *members*.
 *
 * This class has no useful logic; it's just a documentation example.
 *
 * @param T the type of a member in this group.
 * @property name the name of this group.
 * @constructor Creates an empty group.
 */
class Group<T>(val name: String) {
    /**
     * Adds a [member] to this group.
     * @return the new size of the group.
     */
    fun add(member: T): Int { ... }
}
```
3. About string resources. It is not necessary, but you can manually group string, like you are doing now, and then sort comments belong to some group, e.g. select needed lines -> Edit -> Sort lines
<img width="397" alt="image" src="https://github.com/user-attachments/assets/929822b1-1771-4f10-9eb1-eb860b565602">

cValentyna committed:
corrected comments

cValentyna committed:
level 1- corrected silly mistake

EugeneOrlov204 commented on app/src/main/java/com/project/myapp/screens/auth/AuthActivity.kt:
```diff
@@ -0,0 +1,220 @@
+
+    /* Receives name from Email
+     */
+    private fun getName(email: String): String =
+        email
+            .substringBefore("@")
+            .split(".", "_")
+            .joinToString(" ") { it -> it.lowercase().replaceFirstChar { it.uppercaseChar() } }
+}
```
It's a business logic, should be part of the ViewModel

cValentyna replied:
It was moved to the [viewModel](https://github.com/cValentyna/myAndroidProject/pull/11/commits/424fa1694f60903a312870565eb0d4f02883cfe7#:~:text=*/-,fun%20getName(email%3A%20String)%3A%20String%20%3D,%7B%20it%20%2D%3E%20it.lowercase().replaceFirstChar%7B%20it.uppercaseChar()%20%7D%20%7D,-companion%20object%20%7B)

EugeneOrlov204 commented on app/src/main/java/com/project/myapp/screens/auth/AuthActivity.kt:
```diff
@@ -0,0 +1,220 @@
+    }
+
+    private fun isEmailAndPasswordCorrect(): Boolean =
+        viewModel.authState.value.isUserEmailValid &&
+            viewModel.authState.value.isUserPasswordValid &&
+            binding.textInputEditTextAuthEmail.text
+                .toString()
+                .isNotEmpty() &&
+            binding.textInputEditTextAuthPassword.text
+                .toString()
+                .isNotEmpty()
```
why do you need to check edit texts when you already have a authState in the viewmodel?

cValentyna replied:
I needed it because if editText was empty it was considered as correct input data. An it wasn't processed in authState. I will fix it.

EugeneOrlov204 commented on app/src/main/java/com/project/myapp/screens/auth/AuthActivity.kt:
```diff
@@ -0,0 +1,220 @@
+                }
+                textInputLayoutAuthEmail.helperText = getString(R.string.error_email_empty)
+            }
+            if (textInputEditTextAuthPassword.text.toString().isEmpty()
+            ) {
```
Invalid formatting. I suggest you setup the Ktlint: https://pinterest.github.io/ktlint/latest/. We use ktlint in our production product and while having a team of approximately 15 android devs it helps us to maintain good code quality.

cValentyna replied:
I've already use Ktlint. I will try to correct it

EugeneOrlov204 commented on app/src/main/java/com/project/myapp/screens/auth/AuthActivity.kt:
```diff
@@ -0,0 +1,220 @@
+            }
+            if (textInputEditTextAuthPassword.text.toString().isEmpty()
+            ) {
+                viewModel.updateState {
+                    copy(isUserPasswordValid = false)
+                }
```
This looks like a business logic thus shouldn't be on UI layer.

cValentyna replied:
I moved direct call from authState to viewModel [here](https://github.com/cValentyna/myAndroidProject/pull/11/commits/424fa1694f60903a312870565eb0d4f02883cfe7#:~:text=%7D-,fun%20focusEmailUpdate(,%7D,-private%20fun%20getString)
and validate logic [here] (https://github.com/cValentyna/myAndroidProject/pull/11/commits/424fa1694f60903a312870565eb0d4f02883cfe7#diff-2295a611bb9936b9489b1011294a72134f5f861ce9134628200a0eb81359af3bR75:~:text=%3E().getString(int)-,fun%20validatePassword(text%3A%20CharSequence)%3A%20String%3F%20%7B,%7D,-/**)

cValentyna replied:
I moved direct call from authState and validate logic to [viewModel ](https://github.com/cValentyna/myAndroidProject/pull/11/commits/424fa1694f60903a312870565eb0d4f02883cfe7#diff-2295a611bb9936b9489b1011294a72134f5f861ce9134628200a0eb81359af3b)

EugeneOrlov204 commented on app/src/main/java/com/project/myapp/screens/auth/AuthActivity.kt:
```diff
@@ -0,0 +1,220 @@
+            }
+        } else {
+            binding.textInputLayoutAuthPassword.helperText = null
+            viewModel.updateState {
+                copy(isUserPasswordValid = true)
+            }
```
same as below

EugeneOrlov204 commented on app/src/main/java/com/project/myapp/screens/auth/AuthActivity.kt:
```diff
@@ -0,0 +1,220 @@
+        }
+    }
+
+    private fun validatePassword(text: CharSequence?) {
+        // pattern if special_symbol can be present
+        val pattern = Regex("^(?=.*[0-9])(?=.*[a-z])(?=.*[A-Z])[\\w#?!@\$%^&*-]{8,}$")
+        val patternExpectedSymbols = Regex("^[\\w#?!@\$%^&*-]+\$")
+        val patternDigit = Regex("(?=.*[0-9])")
+        val patternLetterLC = Regex("(?=.*[a-z])")
+        val patternLetterUC = Regex("(?=.*[A-Z])")
+        val patternCount = Regex("([\\w#?!@\$%^&*-]{8,})")
+
+        if (!pattern.containsMatchIn(text.toString()) && text.toString().isNotEmpty()) {
+            viewModel.updateState {
+                copy(isUserPasswordValid = false)
+            }
+            when {
+                !patternExpectedSymbols.containsMatchIn(text.toString()) ->
+                    binding.textInputLayoutAuthPassword.helperText =
+                        getString(R.string.error_password_unpredictable_symbols)
+
+                !patternLetterLC.containsMatchIn(text.toString()) ->
+                    binding.textInputLayoutAuthPassword.helperText =
+                        getString(R.string.error_password_lower_case)
+
+                !patternLetterUC.containsMatchIn(text.toString()) ->
+                    binding.textInputLayoutAuthPassword.helperText =
+                        getString(R.string.error_password_upper_case)
+
+                !patternDigit.containsMatchIn(text.toString()) ->
+                    binding.textInputLayoutAuthPassword.helperText =
+                        getString(R.string.error_password_digit)
+
+                !patternCount.containsMatchIn(text.toString()) ->
+                    binding.textInputLayoutAuthPassword.helperText =
+                        getString(R.string.error_password_minimum_characters)
+            }
```
This is definitely a business logic. Please separate from UI.

cValentyna replied:
I moved it to [AuthViewModel](https://github.com/cValentyna/myAndroidProject/pull/11/commits/424fa1694f60903a312870565eb0d4f02883cfe7#diff-2295a611bb9936b9489b1011294a72134f5f861ce9134628200a0eb81359af3b)

EugeneOrlov204 commented on app/src/main/java/com/project/myapp/screens/auth/AuthActivity.kt:
```diff
@@ -0,0 +1,220 @@
+        }
+    }
+
+    /*
+    If register button was pressed or textInputEditText was focused and lost focus,
+    error will be shown if input text did not pass validation
+    when user is correcting:
+    email - error will not be shown
+    password - error will be shown until password is correct
+     */
```
I advise you to use [Kdoc](https://kotlinlang.org/docs/kotlin-doc.html#block-tags) for better comments. Also, some tools can convert your Kdocs to a nice looking documentation.

EugeneOrlov204 commented on app/src/main/java/com/project/myapp/screens/auth/AuthActivity.kt:
```diff
@@ -0,0 +1,220 @@
+    private fun setOnClickListener() {
+        binding.apply {
+            buttonAuthRegister.setOnClickListener {
+                validateEmail(textInputEditTextAuthEmail.text)
+                validatePassword(textInputEditTextAuthPassword.text)
+                if (isEmailAndPasswordCorrect()) {
```
it's odd to see that you validate email and password and then check their validity. I would rather have some link this:
```suggestion
                if (isEmailValid() && isPasswordValid()) {
```

EugeneOrlov204 replied:
Also, it makes sense to keep in setOnClickListener just setting listeners and move all the click events to separate functions.

cValentyna replied:
I moved this logic to separated functions [ba705ab](https://github.com/cValentyna/myAndroidProject/pull/11/commits/ba705abf1603225de65a864cf10707dff8327df4), but after next movements it was a little bit changed in [424fa16](https://github.com/cValentyna/myAndroidProject/pull/11/commits/424fa1694f60903a312870565eb0d4f02883cfe7)

EugeneOrlov204 commented on app/src/main/java/com/project/myapp/screens/auth/AuthState.kt:
```diff
@@ -0,0 +1,14 @@
+/*
+ Class where information about the state is stored
+ */
+data class AuthState(
+    var isUserEmailValid: Boolean = false,
+    var isUserPasswordValid: Boolean = false,
+    var wasFocusEmail: Boolean = false,
+    var firstFocusEmail: Boolean = false,
+    var wasFocusPassword: Boolean = false,
+    var firstFocusPassword: Boolean = false,
+    var wasRegisterButtonClicked: Boolean = false,
+)
```
Please read about single responsibility. This is a violation of this principle: Auth state keeps track of the following:  
- Email validity
- Password validity
- Several different focuses (which I suspect might be a focus of EditText which should be part of the UI, not the EditText)
- Register button clickable state

Usually, you would create an enum or a sealed class to keep track of your authorization status, such as SUCCESS, INVALID_EMAIL, INVALID_PASSWORD, NETWORK_ERROR etc. But this requires some time to think and create scalable and reusable architecture.

kogorbach replied:
agreed, I think MVI principles would have a guideline on this approach, @EugeneOrlov204 can you confirm?

cValentyna replied:
I tried to change [AuthState class](https://github.com/cValentyna/myAndroidProject/pull/11/commits/984ad6ae893840e1ca29a5c223a7b002e4004efd?diff=split&w=0) as sealed class, therefore [viewModel](https://github.com/cValentyna/myAndroidProject/pull/11/commits/fc6731b6ab19ef4edce4de56cd758e621cf0f9f8) was changed, and observers were added in [AuthActivity](https://github.com/cValentyna/myAndroidProject/pull/11/commits/c06bd793abf380069d740aa5dcd640d619394f32?diff=split&w=0#).

EugeneOrlov204 replied:
> agreed, I think MVI principles would have a guideline on this approach, @EugeneOrlov204 can you confirm?

Not sure I understand what you are saying...

EugeneOrlov204 commented on app/src/main/java/com/project/myapp/screens/auth/AuthViewModel.kt:
```diff
@@ -0,0 +1,19 @@
+  ViewModel for saving AuthActivity state
+ */
+
+class AuthViewModel : ViewModel() {
+    private val _authState = MutableStateFlow(AuthState())
+    val authState: StateFlow<AuthState> get() = _authState.asStateFlow()
+
+    fun updateState(reducer: AuthState.() -> AuthState) {
+        _authState.update(reducer)
+    }
+}
```
I am happy to see that you aware about ViewModels and use them, but please try to separate UI from business logic as was suggested in comments above.

kogorbach replied:
I think this is actually a valid approach tbh

YevheniiOrlovIntellias replied:
Please look at the AuthState implementation. I agree that in general this view model looks ok, but the way it is used is not

kogorbach replied:
Yeah the implementation is questionable, agree here, but overall having a UI state in viewmodel doesn't seem to be a bad or uncommon practice

EugeneOrlov204 commented on app/src/main/java/com/project/myapp/screens/main/MainActivity.kt:
```diff
@@ -13,16 +16,27 @@ class MainActivity : AppCompatActivity() {
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContentView(binding.root)
+        getName()
```
"get" suggests that this method should return a value:
![image](https://github.com/user-attachments/assets/df93a15a-04f8-473a-ae15-5089d4fd5b14)
https://en.wikibooks.org/wiki/Object_Oriented_Programming/Getters_and_Setters

cValentyna replied:
It was fixed [0e586d4cbf390426ae6c5878f4f7109c691f94aa](https://github.com/cValentyna/myAndroidProject/pull/11/commits/0e586d4cbf390426ae6c5878f4f7109c691f94aa)

EugeneOrlov204 commented on app/src/main/java/com/project/myapp/screens/main/MainActivity.kt:
```diff
@@ -13,16 +16,27 @@ class MainActivity : AppCompatActivity() {
             buttonMainLogout.setOnClickListener {
                 val intent = Intent(this@MainActivity, AuthActivity::class.java)
-                startActivity(intent)
+                val option =
+                    ActivityOptionsCompat.makeCustomAnimation(
+                        this@MainActivity,
+                        R.anim.slide_in_right,
+                        R.anim.slide_out_right,
+                    )
+                startActivity(intent, option.toBundle())
```
move this logic to a separate function

cValentyna replied:
It was changed [0e586d4cbf390426ae6c5878f4f7109c691f94aa](https://github.com/cValentyna/myAndroidProject/pull/11/commits/0e586d4cbf390426ae6c5878f4f7109c691f94aa)

kogorbach replied:
@EugeneOrlov204 please approve for resolve so we can clean up conversation a bit

EugeneOrlov204 replied:
@kogorbach I am not sure how to do so, it seems that either I am stupid or don't have rights for that :)

kogorbach commented on app/src/main/java/com/project/myapp/screens/auth/AuthActivity.kt:
```diff
@@ -0,0 +1,220 @@
+                    viewModel.updateState {
+                        copy(wasRegisterButtonClicked = true)
+                    }
+                    Toast
```
I'd create an extension funcion on Context for making toasts, you can check our 'ext' channel in discord for this and other extensions

cValentyna replied:
I added [extensionFunction](https://github.com/cValentyna/myAndroidProject/pull/11/commits/ba705abf1603225de65a864cf10707dff8327df4#diff-ba28775a683a4f8a67bff9a9aa7044df9cdb907bb4cb560227644fd76b17521b)

kogorbach commented on app/src/main/java/com/project/myapp/screens/auth/AuthActivity.kt:
```diff
@@ -0,0 +1,220 @@
+                            R.anim.slide_in_left,
+                            R.anim.slide_out_left,
+                        )
+                    intent.putExtra("userName", userName)
```
hardcoded key string is dangerous for typos, better store them in a separate file or object

cValentyna replied:
I changed it.

kogorbach commented on app/src/main/java/com/project/myapp/screens/auth/AuthActivity.kt:
```diff
@@ -0,0 +1,220 @@
+    private fun setFocusListener() {
+        binding.apply {
+            textInputEditTextAuthEmail.setOnFocusChangeListener { v, hasFocus ->
+                if (hasFocus) {
```
kudos for doing the focus change stuff

kogorbach commented on app/src/main/java/com/project/myapp/screens/auth/AuthActivity.kt:
```diff
@@ -0,0 +1,220 @@
+                    viewModel.updateState {
+                        copy(firstFocusPassword = true)
+                    }
+                } else if (!v.isFocused && viewModel.authState.value.firstFocusPassword) {
```
normally you wouldn't access values of viewmodels flows/states/liveData from UI layer, I'd suggest reviewing if you can move this logic to the viewmodel

cValentyna replied:
I moved validation and direct call authState to [viewModel](https://github.com/cValentyna/myAndroidProject/pull/11/commits/424fa1694f60903a312870565eb0d4f02883cfe7#diff-2295a611bb9936b9489b1011294a72134f5f861ce9134628200a0eb81359af3b)

kogorbach commented on app/src/main/java/com/project/myapp/screens/auth/AuthActivity.kt:
```diff
@@ -0,0 +1,220 @@
+        }
+    }
+
+    private fun validateEmail(text: CharSequence?) {
```
consider doing the validation process in a viewmodel or a custom validator class, you can return the R.string id from there and apply it in the activity. Hope it makes sense, feel free to reach out for clarifications

cValentyna replied:
Validation was moved to [viewModel](https://github.com/cValentyna/myAndroidProject/pull/11/commits/424fa1694f60903a312870565eb0d4f02883cfe7#diff-2295a611bb9936b9489b1011294a72134f5f861ce9134628200a0eb81359af3b)

kogorbach commented on app/src/main/java/com/project/myapp/screens/auth/AuthActivity.kt:
```diff
@@ -0,0 +1,220 @@
+        val pattern = Regex("^(?=.*[0-9])(?=.*[a-z])(?=.*[A-Z])[\\w#?!@\$%^&*-]{8,}$")
+        val patternExpectedSymbols = Regex("^[\\w#?!@\$%^&*-]+\$")
+        val patternDigit = Regex("(?=.*[0-9])")
+        val patternLetterLC = Regex("(?=.*[a-z])")
```
could be just me, but it's not too obvious that LC and UC are lower case and upper case, also I'm not a big fan of regex so I'd probably avoid them completely. I will text you an example how I managed some text validations recently, just to share ideas

cValentyna replied:
It was changed [here](https://github.com/cValentyna/myAndroidProject/pull/11/commits/424fa1694f60903a312870565eb0d4f02883cfe7#diff-2295a611bb9936b9489b1011294a72134f5f861ce9134628200a0eb81359af3bR75:~:text=fun%20validatePassword(text%3A%20CharSequence)%3A,return%20null)   
But i didn't follow your example.

kogorbach commented on app/src/main/java/com/project/myapp/screens/auth/AuthActivity.kt:
```diff
@@ -0,0 +1,220 @@
+        }
+    }
+
+    private fun isEmailAndPasswordCorrect(): Boolean =
```
Having 'and' in function's name sugggests doing several operations in it, which we probably always want to avoid

kogorbach commented on app/src/main/java/com/project/myapp/screens/main/MainActivity.kt:
```diff
@@ -13,16 +16,27 @@ class MainActivity : AppCompatActivity() {
+
+    private fun getName() {
+        binding.textViewMainUserName.text = intent.getStringExtra("userName")
+    }
```
hardcoded string key, mentioned earlier, just pinging it here as well to notice

cValentyna replied:
It was changed [0e586d4cbf390426ae6c5878f4f7109c691f94aa](https://github.com/cValentyna/myAndroidProject/pull/11/commits/0e586d4cbf390426ae6c5878f4f7109c691f94aa)

kogorbach replied:
nice, please resolve here

kogorbach commented on app/src/main/java/com/project/myapp/screens/auth/AuthViewModel.kt:
```diff
@@ -0,0 +1,19 @@
+    private val _authState = MutableStateFlow(AuthState())
+    val authState: StateFlow<AuthState> get() = _authState.asStateFlow()
+
+    fun updateState(reducer: AuthState.() -> AuthState) {
```
I had to look up what this AuthState.() was, apparently this is called extension function receiver, and looks to me like a very high level thing, great job!

cValentyna replied:
It is Stanislav's idea and suggestion, because I tried to listen to focuses and other authStates without viewModel and a separated class.

cValentyna committed:
level1 -  renamed fun, changed hardcoded string, moved logOut logic to separate fun

cValentyna committed:
level1 -  changed hardcoded string, moved Register button logic to separate functions, added extension function for toast

cValentyna committed:
level1 -  moved validation and ui state logic to viewmodel

kogorbach commented on app/src/main/java/com/project/myapp/ExFunctions.kt:
```diff
@@ -0,0 +1,10 @@
+import android.widget.Toast
+import androidx.appcompat.app.AppCompatActivity
+
+object ExtensionUtils {
```
you don't need an object to create extension functions, it is usually a plain file
e.g.
<img width="958" alt="image" src="https://github.com/user-attachments/assets/7e60ebce-23d0-418e-a0e4-f63a9085b6a4">

cValentyna replied:
It was [changed ](https://github.com/cValentyna/myAndroidProject/pull/11/commits/372427c33fa5983eb0e53a1c5f21c53fb2bcc8d3#diff-ba28775a683a4f8a67bff9a9aa7044df9cdb907bb4cb560227644fd76b17521b)

kogorbach commented on app/src/main/java/com/project/myapp/screens/auth/AuthViewModel.kt:
```diff
@@ -1,19 +1,156 @@
+        }
+    }
+
+    private fun getString(int: Int): String = getApplication<Application>().getString(int)
```
you shouldn't access context and resources in a viewmodel, better to return the resource id instead, and then access it from activity or fragment

cValentyna replied:
I changed [it](https://github.com/cValentyna/myAndroidProject/pull/11/commits/372427c33fa5983eb0e53a1c5f21c53fb2bcc8d3#diff-2295a611bb9936b9489b1011294a72134f5f861ce9134628200a0eb81359af3b).
I hope that's what you meant.

kogorbach replied:
yes, beautifuly done, you might want to also annotate return type of those functions with @StringRes so it's more clear that it's not just an Int

cValentyna committed:
level1 - removed unnecessary object around extension function; returns resource ID instead of using context

cValentyna committed:
level1 - added annotation StringRes

cValentyna committed:
level1 - renamed class HolderKeys.kt to KeysHolder. A little correction according to Ktlint

cValentyna committed:
level1 - renamed class HolderKeys.kt to KeysHolder. A little correction according to Ktlint

cValentyna committed:
level1 - rewrote AuthState as sealed class

cValentyna committed:
level1 - changed AuthViewModel

cValentyna committed:
level1 - changed AuthActivity

cValentyna committed:
level1 - added @StringRes annotation

kogorbach commented on app/src/main/java/com/project/myapp/ExFunctions.kt:
```diff
@@ -0,0 +1,8 @@
+import android.widget.Toast
+import androidx.appcompat.app.AppCompatActivity
+
+    fun AppCompatActivity.toast(message: String) {
```
imo, an extension on Context would be more flexible as you would also be able to call it from activities, fragments, dialogs, etc

kogorbach replied:
also a more popular approach would be to have an 'ext' or 'extension' folder, and then separate files for each class you create an extension for, in this case, it would be ext.Activity, or ext.Context, if you change to context extension

cValentyna replied:
It was changed [here](https://github.com/cValentyna/myAndroidProject/pull/11/commits/31992ff1e7f14b2cc1c68649897dee36e23ebd7b)

kogorbach commented on app/src/main/java/com/project/myapp/screens/auth/AuthActivity.kt:
```diff
@@ -0,0 +1,129 @@
+    private fun setObservers() {
+        viewModel.passwordState.observe(this) { state ->
+            binding.textInputLayoutAuthPassword.error =
+                when (state) {
```
loving this, it's important to keep exhaustive when blocks, like this one, not to forget to adjust those if a new Enum/sealed class value is introduced

kogorbach commented on app/src/main/java/com/project/myapp/screens/auth/AuthActivity.kt:
```diff
@@ -0,0 +1,129 @@
+        viewModel.credentialsState.observe(this) { state ->
+            when (state) {
+                is AuthState.Error ->
+                    this.toast(state.message?.let { getString(it) }.toString())
```
i think it can work without 'this.'

cValentyna replied:
corrected it

EugeneOrlov204 commented on app/src/main/java/com/project/myapp/screens/auth/AuthState.kt:
```diff
@@ -0,0 +1,25 @@
+
+    sealed class EmailState {
+        data object Valid : EmailState()
+        data class Error(val message: Int?) : EmailState()
+        data class Empty(val message: Int?) : EmailState()
+        data object InvisibleError : EmailState()
+    }
+
+    sealed class PasswordState {
+        data object Valid : PasswordState()
+        data class Error(val message: Int?) : PasswordState()
+        data class Empty(val message: Int?) : PasswordState()
+        data object InvisibleError : PasswordState()
```
In my humble opinion, it's better to not store any resources in these classes at all but rather to have a separate state for each possible error and then, based on the error state, display appropriate text message

cValentyna replied:
I changed it [here](https://github.com/cValentyna/myAndroidProject/pull/11/commits/76d257965af4c89ba8b9a712dbaa9afced8bfd79?diff=split&w=0#diff-ff6b82efc6079a290f10a2da121531c0d479abdcf6b2ac9e57b8500eb7c78e49)

EugeneOrlov204 commented on app/src/main/java/com/project/myapp/screens/auth/AuthActivity.kt:
```diff
@@ -0,0 +1,129 @@
+        viewModel.passwordState.observe(this) { state ->
+            binding.textInputLayoutAuthPassword.error =
+                when (state) {
+                    is AuthState.PasswordState.Error -> state.message?.let { getString(it) }
+                    is AuthState.PasswordState.Empty -> state.message?.let { getString(it) }
+                    is AuthState.PasswordState.InvisibleError,
+                    is AuthState.PasswordState.Valid -> null
+                }
+        }
+
+        viewModel.emailState.observe(this) { state ->
+            binding.textInputLayoutAuthEmail.error =
+                when (state) {
+                    is AuthState.EmailState.Error -> state.message?.let { getString(it) }
+                    is AuthState.EmailState.Empty -> state.message?.let { getString(it) }
+                    is AuthState.EmailState.InvisibleError,
+                    is AuthState.EmailState.Valid
+                    -> null
+                }
+        }
```
To me it looks like code duplication

EugeneOrlov204 commented on app/src/main/java/com/project/myapp/screens/auth/AuthActivity.kt:
```diff
@@ -0,0 +1,129 @@
+
+
+    private fun goToNextActivity() {
+        val userName = viewModel.getName(binding.textInputEditTextAuthEmail.text.toString())
+        val intent = Intent(this@AuthActivity, MainActivity::class.java)
+        val option =
+            ActivityOptionsCompat.makeCustomAnimation(
+                this@AuthActivity,
+                R.anim.slide_in_left,
+                R.anim.slide_out_left,
+            )
+        intent.putExtra(USER_NAME_KEY, userName)
+        startActivity(intent, option.toBundle())
+        finish()
```
also, you can do something like:
Intent().apply {
putExtra() 
doAdditionalActionOnTheObject()
}

cValentyna replied:
It was changed [here](https://github.com/cValentyna/myAndroidProject/pull/11/commits/8601c9565859f01af4e04b211da5cb3b53cc559d#diff-003477427236e4babf86151628b05bb3917f7fdd61b2351b73d6b5631e5d588a:~:text=val%20userName%20%3D%20viewModel.parseName,startActivity(intent%2C%20animation.toBundle()))

EugeneOrlov204 commented on app/src/main/java/com/project/myapp/screens/auth/AuthViewModel.kt:
```diff
@@ -0,0 +1,178 @@
+
+    fun validateEmail(email: String) {
+        val pattern = Patterns.EMAIL_ADDRESS.matcher(email).matches()
+        if (!pattern && email.isNotEmpty()){
```
I haven't done this myself on my pet projects, but consider setting up a GitHub CI with a linter on it. I think it would be a good point during interviews

EugeneOrlov204 replied:
I haven't done this myself on my pet projects, but consider setting up a GitHub CI with a linter on it. I think it would be a good point during interviews

EugeneOrlov204 commented on app/src/main/java/com/project/myapp/screens/auth/AuthViewModel.kt:
```diff
@@ -0,0 +1,178 @@
+     * Gets an email error when  email is invalid
+     */
+
+    private fun getEmailError(): Int {
+        return R.string.error_e_mail_address
+    }
```
ViewModel should not be aware of string resources like this.

kogorbach replied:
why not? it's not accessing it, only returning the resource id. The question remains though is that this method doesn't do anything, could just as well call the respective string resource directly instead of calling the viewmodel method

EugeneOrlov204 replied:
"Email error" in my opinion is too vague of a description. It can be anything from empty email, invalid email to email that already exists errors that I would rather rely on enums and assign appropriate text on the UI side. Also, this solution would be beneficial since you could share the state across multiple screens which would react differently to the same error message (but in this particular case it's unlikely)

cValentyna replied:
I changed this and the same methods [here](https://github.com/cValentyna/myAndroidProject/pull/11/commits/76d257965af4c89ba8b9a712dbaa9afced8bfd79?diff=split&w=0#diff-2295a611bb9936b9489b1011294a72134f5f861ce9134628200a0eb81359af3bR27-R39). And thereform [activity](https://github.com/cValentyna/myAndroidProject/pull/11/commits/76d257965af4c89ba8b9a712dbaa9afced8bfd79?diff=split&w=0#diff-003477427236e4babf86151628b05bb3917f7fdd61b2351b73d6b5631e5d588a) was changed.

kogorbach replied:
LGTM here

EugeneOrlov204 commented on app/src/main/java/com/project/myapp/screens/auth/AuthViewModel.kt:
```diff
@@ -0,0 +1,178 @@
+        } else {
+            _passwordState.value =AuthState.PasswordState.Valid
+        }
+    }
```
Your auth validation logic has grown to be complex, consider moving it to a separate class. It would be ideal if you set up a DI framework and inject that class using it.

cValentyna replied:
I tried to do it with "koin". And it even works, but I'm not sure I understand all of it. I`ve not pusshed these changes.

kogorbach replied:
Using a DI is a junior level, understanding it is a strong middle level or higher, it's fine, as long as you see the pattern in your head and can duplicate injections, also wouldn't be a bad idea to check Hilt

kogorbach commented on app/src/main/java/com/project/myapp/screens/auth/AuthViewModel.kt:
```diff
@@ -0,0 +1,178 @@
+     */
+
+    fun validateEmail(email: String) {
+        val pattern = Patterns.EMAIL_ADDRESS.matcher(email).matches()
```
doesn't that pattern cover isEmpty case?

EugeneOrlov204 commented on app/src/main/java/com/project/myapp/screens/auth/AuthViewModel.kt:
```diff
@@ -0,0 +1,178 @@
+            .substringBefore("@")
+            .split(".", "_")
+            .joinToString(" ") { it -> it.lowercase().replaceFirstChar { it.uppercaseChar() } }
+
```
for now it's ok but I would move it to another class/object

EugeneOrlov204 commented on app/src/main/java/com/project/myapp/screens/auth/AuthViewModel.kt:
```diff
@@ -0,0 +1,178 @@
+     * when user clicked authorization button
+     */
+
+    private fun checkEmailByClick(email:String) {
+        if (email.isEmpty()) {
+            _emailState.value = AuthState.EmailState.Empty(R.string.error_email_empty)
+        } else {
+            validateEmail(email)
+        }
+    }
+
+    /**
+     * Validates
+     * @param password
+     * when user clicked authorization button
+     */
+    private fun checkPasswordByClick(password: String) {
+        if (password.isEmpty()) {
+            _passwordState.value = AuthState.PasswordState.Empty(R.string.error_password_empty)
+        } else {
+            validatePassword(password)
+        }
+    }
```
I think methods validatePassword and validateEmail should have validated for the empty field case thus I don't see a point for having an additional method like this.

cValentyna replied:
these methods were removed and their functions was moved [here](https://github.com/cValentyna/myAndroidProject/pull/11/commits/76d257965af4c89ba8b9a712dbaa9afced8bfd79?diff=split&w=0#diff-2295a611bb9936b9489b1011294a72134f5f861ce9134628200a0eb81359af3bR27-R39) and for password - below

kogorbach commented on app/src/main/java/com/project/myapp/screens/auth/AuthActivity.kt:
```diff
@@ -0,0 +1,129 @@
+
+
+    private fun goToNextActivity() {
+        val userName = viewModel.getName(binding.textInputEditTextAuthEmail.text.toString())
```
it's not that critical, but 'get' sounds to me like fetching from a data source like network API, what we do in viewmodel sounds more like parsing

cValentyna replied:
it was changed

kogorbach commented on app/src/main/java/com/project/myapp/screens/auth/AuthActivity.kt:
```diff
@@ -0,0 +1,129 @@
+    private fun goToNextActivity() {
+        val userName = viewModel.getName(binding.textInputEditTextAuthEmail.text.toString())
+        val intent = Intent(this@AuthActivity, MainActivity::class.java)
+        val option =
```
it looks to me like an animation, why option?

cValentyna replied:
I think because ActivityOptionsCompat is used here. I found an example of implementing this animation and used it.. I changed the name.

kogorbach commented on app/src/main/java/com/project/myapp/screens/auth/AuthState.kt:
```diff
@@ -0,0 +1,25 @@
+
+    sealed class EmailState {
+        data object Valid : EmailState()
+        data class Error(val message: Int?) : EmailState()
```
should message be nullable here and in the rest of error states?

cValentyna replied:
It could be needed in previous variant of code because my validation functions could return null in cases when checked data was emty(now empty data processes together with other validation). but I changed [AuthState](https://github.com/cValentyna/myAndroidProject/pull/11/commits/76d257965af4c89ba8b9a712dbaa9afced8bfd79?diff=split&w=0#diff-ff6b82efc6079a290f10a2da121531c0d479abdcf6b2ac9e57b8500eb7c78e49) class and Error() does not receive message now.

kogorbach commented on app/src/main/java/com/project/myapp/screens/auth/AuthViewModel.kt:
```diff
@@ -0,0 +1,178 @@
+     * Variables to track the states
+     */
+    private val _credentialsState = MutableLiveData<AuthState>()
+    val credentialsState: LiveData<AuthState> get() = _credentialsState
```
not a mistake, just a recommendation to try out flow, I've seen much more flows than live datas recently, would be useful

EugeneOrlov204 replied:
LiveData is a rare bird these days :)

cValentyna replied:
changes to flow are [here](https://github.com/cValentyna/myAndroidProject/pull/11/commits/87658ae873dcdb3282cde086ef092f0885d3892d)

kogorbach commented on app/src/main/java/com/project/myapp/screens/auth/AuthViewModel.kt:
```diff
@@ -0,0 +1,178 @@
+    fun validatePassword(password: String) {
+        val isFailed =
+            password.isNotEmpty() && (
+                    !password.validSigns() ||
```
the same methods are called in getError(), so you could call it right away, and rely on it returning null to determine valid password

cValentyna replied:
these methods were removed and their functions was moved [here](https://github.com/cValentyna/myAndroidProject/pull/11/commits/76d257965af4c89ba8b9a712dbaa9afced8bfd79?diff=split&w=0#diff-2295a611bb9936b9489b1011294a72134f5f861ce9134628200a0eb81359af3bR27-R39) for email and for password - below

kogorbach commented on app/src/main/java/com/project/myapp/screens/auth/AuthViewModel.kt:
```diff
@@ -0,0 +1,178 @@
+            !text.validSigns() ->
+                R.string.error_password_unpredictable_symbols
+
+            text.onlyLetters() ->
```
ghwfeio@#@
not only letters, still missing a digit, similar cases for other checks

kogorbach replied:
oh, you seem to forbid special characters here, if that's the case, you can resolve and ignore this comment, my bad

kogorbach reviewed (changes requested):
changes requested, overall comment: you are overdoing comments, variables and namings are generally very descriptive and precise, that is enough to understand what your code does. Don't comment on **what** it does, comment **why** it does so, if needed of course. One case where I would keep the comment is your workaround with invisible error where you pause validation

cValentyna committed:
level1 - changed LiveData to StateFlow

cValentyna committed:
level1 - AuthStates: added states for each possible error, removed received strings from error classes.
AuthViewModel : rewrote  email-password validations methods, removed unnecessary methods.
AuthActivity:  changed funSetObservers according to previous changes, renamed it.
strings.xml: renamed string name for email error

cValentyna committed:
level1 - moved extension function to the ext folder.
changed extension to Context

cValentyna committed:
level1 - renamed val "option" to "animation" for animation between activities

cValentyna committed:
level1 - changed according to ktlint and other little changes

EugeneOrlov204 commented on app/src/main/java/com/project/myapp/screens/auth/AuthActivity.kt:
```diff
@@ -83,26 +83,30 @@ class AuthActivity : AppCompatActivity() {
                         is AuthState.PasswordState.InvisibleError,
                         is AuthState.PasswordState.Valid,
                         is AuthState.PasswordState.Initial,
                         -> null
```
why not `else -> null`?

cValentyna replied:
when I added new states, this view was clearer to me. You are right, your solution is better. It was changed [here](https://github.com/cValentyna/myAndroidProject/pull/11/commits/1960d2a943f4056ef58550c3ade1c82e7125da08#diff-003477427236e4babf86151628b05bb3917f7fdd61b2351b73d6b5631e5d588a).

EugeneOrlov204 commented on app/src/main/java/com/project/myapp/screens/auth/AuthViewModel.kt:
```diff
@@ -26,77 +24,57 @@ class AuthViewModel : ViewModel() {
 
     /**
-     * Validates email when email is not empty  and changes _emailState.value
+     * Validates email, changes _emailState.value as AuthState.EmailState
+     * which depends on filled data
```
Comment should not be obvious and duplicate what is written in code. Usually, you would comment the parts of the code that are hard to understand without additional explanations.

cValentyna replied:
It was a little bit changed [here](https://github.com/cValentyna/myAndroidProject/pull/11/commits/1960d2a943f4056ef58550c3ade1c82e7125da08)

cValentyna committed:
level1 - corrected comments, removed duplicate code for similar states

EugeneOrlov204 commented on app/src/main/java/com/project/myapp/screens/auth/AuthState.kt:
```diff
@@ -0,0 +1,43 @@
+package com.project.myapp.screens.auth
+
+/**
+ * Class where ui state is stored
+ */
```
redundant comment

cValentyna replied:
Comment was removed.

EugeneOrlov204 commented on app/src/main/java/com/project/myapp/screens/auth/AuthState.kt:
```diff
@@ -0,0 +1,43 @@
+    data object Error : AuthState()
+
+    sealed class EmailState {
+        data object Initial : EmailState()
```
Maybe, you could make EmailState extend the AuthState and use the latter's Initial state? I mean, you structured the code as if it would be an inheritance but it does not look like it

cValentyna replied:
EmailState and PasswordState [was extended](https://github.com/cValentyna/myAndroidProject/pull/11/commits/73cede8bf68e30a6647bd53b47bc91a1908b2514) AuthState, but this part :

>  use the latter's Initial state

wasn't clear to me. EmailState and PasswordState changes separately and AuthState depends on both (that`s why EmailState and PasswordState are located inside AuthState).

EugeneOrlov204 commented on app/src/main/java/com/project/myapp/screens/auth/AuthViewModel.kt:
```diff
@@ -0,0 +1,121 @@
+
+    fun validateEmail(email: String) {
+        val pattern = Patterns.EMAIL_ADDRESS.matcher(email).matches()
+        if (!pattern && email.isNotEmpty()) {
+            _emailState.value = AuthState.EmailState.Error
+        } else if (email.isEmpty()) {
+            _emailState.value = AuthState.EmailState.ErrorEmpty
+        } else if (pattern) {
+            _emailState.value = AuthState.EmailState.Valid
+        }
```
what do you think about this?
```suggestion
      _emailState.value = if (email.isEmpty()) {
             AuthState.EmailState.ErrorEmpty
        } else if (!pattern) {
             AuthState.EmailState.Error
        } else {
             AuthState.EmailState.Valid
        }
```

cValentyna replied:
Your solution is definitely better and clearer. I changed [it](https://github.com/cValentyna/myAndroidProject/pull/11/commits/08ab598199580641ae3d9cca49db7081f02819a0)

cValentyna committed:
level1 - changed  the view fun validateEmail()

cValentyna committed:
level1 - removed unnecessary comment; EmailState and PasswordState extends AuthState
